content/blog/cran-checks-compiled-code/index.Rmd

```diff
@@ -0,0 +1,225 @@
+
+After making these changes, the package should install without trouble on R 3.6 and above. It may not build on R <= 3.5 on systems with very old compilers, though it is likely that the vast majority of users will have a newer version of R and/or have recent enough compilers. If you want to be confident that your package will be installable on R 3.5 and below with old compilers, there are several options; we offer two of the simplest approaches here:
+
+-   You can use a configure script at the top level of the package, and have it add `CXX_STD=CXX11` for R 3.5 and below. An example (unmerged) [pull request to the readxl](https://github.com/tidyverse/readxl/pull/722/files) package demonstrates this approach. You will also need to add `Biarch: true` in your DESCRIPTION file. This appears to be the approach preferred by CRAN.
+-   For users with R <= 3.5 on a system with an older compiler, package authors can instruct users to edit their `~/.R/Makevars` file to include this line: `CXX_STD=CXX11`.
+
+The tidyverse has a [policy of supporting four previous versions](https://www.tidyverse.org/blog/2019/04/r-version-support/) of R. Currently that includes R 3.5, but with the upcoming release of R 4.3 (which should be this Spring some time) the minimum version we will support is R 3.6. As we won't be supporting R 3.5 in the near future, you should not feel pressured to either unless you have a compelling reason.
+
+## WARNING regarding the use of <code>sprintf()</code> in C/C++
+
+Another recent change in CRAN checks on R-devel that authors might encounter is the disallowing of the use of the C functions <code>sprintf()</code> and `vsprintf()`. `R CMD check` on R-devel may throw warnings that look something like this:
+
+```
+checking compiled code ... WARNING
+File 'fs/libs/fs.so':
+  Found 'sprintf', possibly from 'sprintf' (C)
+    Object: 'file.o'
+Compiled code should not call entry points which might 
+terminate R nor write to stdout/stderr instead of to the 
+console, nor use Fortran I/O nor system RNGs nor [v]sprintf.
+See 'Writing portable packages' in the 'Writing R Extensions' manual.
+```
+
+According to the [NEWS for R-devel](https://developer.r-project.org/blosxom.cgi/R-devel/NEWS/2022/12/24#n2022-12-24) (which will be R 4.3):
+
+> The use of sprintf and vsprintf from C/C++ has been deprecated in macOS 13 and is a known security risk. `R CMD check` now reports (on all platforms) if their use is found in compiled code: replace by snprintf or vsnprintf respectively.
+
+These are considered to be a security risk because they potentially allow [buffer overflows](https://en.wikipedia.org/wiki/Buffer_overflow) that write more bytes than are available in the output buffer. This is a risk if the text that is being passed to <code>sprintf()</code> comes from an uncontrolled source.
+
+Here is a very simple example:
+
+```{r}
+library(cpp11)
+
+cpp_function('
+  int say_height(int height) {
```

Comment:
Even here I think capturing the `n` is a bit unusual. It's probably best practice, but I think few people do it, and you get to the heart of the problem faster without it.

Reply:
Fair enough. I was doing it to demonstrate the return value vs the `size` parameter - I've seen places where people capture it and compare it to the expected `size` and then condition on that...

Reply:
If you're ok with it @hadley I'd rather keep the `n` - I think it helps to demonstrate the slightly different return values of `sprintf()` vs `snprintf()`.

Reply:
I don't have strong feelings about it so go ahead.

Reply:
Thanks!